### PR TITLE
(refactor) Patient Edit/Update

### LINKED ISSUE #219: Patient Update - 2.x Review
This issue tracks the current state of the patient update 2.x review - once the issue has been completed it can be submitted for `PRODUCTION READY` review. 
(Detailed progress is represented visually [here](https://docs.google.com/spreadsheets/d/1sAEIbcAY2x-EVPCTKmMacn5LL3mtw9x7GM0P-eyJxjs/edit?usp=sharing))
### Essential updates
- [x] Resolve bug introduced into `master` branch causing the patient update module to not load 
### Form
- [ ] Make use of ng-messages to provide form feedback 
- [ ] Latest form validation standards 
### Tests
- [ ] Implement end to end tests to cover editing patient groups, debtor groups and patient attributes.

### PR DESCRIPTION
**Pull request by jniles**
This commit refactors the patient edit state to use the current 2.x best practices.
1. `bhLoadingButton` is introduced to handle form submission.
2. `uib-datepicker` is used for the inline date.
3. Date escaping is done on the server, not the client.
4. Font awesome preferred over glyphicons

This commit refactors the patient edit modals to use the 2.x best practices.  These include:
1. `bhLoadingButton` for all submit buttons.
2. Proper HTTP error handling with `Notify`.

With these improvements, patient edit passes the review criteria and can now be considered completed.

Closes #219.

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

File: client/src/partials/patients/edit/updateDebtorGroup.modal.js [hunk 1]
```diff
@@ -1,60 +1,69 @@
...
-      if (debtorGroup.uuid === uuid) { 
-        name = debtorGroup.name;
-        return true;
+    var groups = viewModel.debtorGroups;
+    var i = groups.length;
+
+    while (i--) {
+      if (groups[i].uuid === uuid) {
+        name = groups[i].name;
+        break;
```
Comment (mbayopanda) [inline client/src/partials/patients/edit/updateDebtorGroup.modal.js:29]: 👍 very nice

File: test/end-to-end/patient/update.spec.js [hunk 2]
```diff
@@ -40,21 +38,16 @@ describe('Patient Edit', function () {
...
     element(by.css('[data-update-group-debtor]')).click();
-
-    let select = element(by.model('UpdateDebtorGroupCtrl.debtor_group_uuid'));
-    let option = select.element(by.cssContainingText('option', 'Second Test Debtor Group'));
-    option.click();
-
-    element(by.css('[data-confirm-group]')).click();
-
+    FU.select('UpdateDebtorGroupCtrl.debtor_group_uuid','Second Test Debtor Group');
+    FU.modal.submit();
```
Comment (mbayopanda) [inline test/end-to-end/patient/update.spec.js:42]: 👍

File: client/src/partials/patients/edit/updatePatientGroups.modal.js [hunk 3]
```diff
@@ -1,67 +1,68 @@
...
         closeModal();
       });
   };
 
   // TODO Refactor - use stores?
-  function fetchGroupObject(uuid) { 
+  function fetchGroupObject(uuid) {
     var groupObject;
-    viewModel.patientGroups.some(function (patientGroup) { 
+    viewModel.patientGroups.some(function (patientGroup) {
```
Comment (mbayopanda) [inline client/src/partials/patients/edit/updatePatientGroups.modal.js]: The `while(i--)` syntax cannot be applied here ? i think it is more clear, what do you think about that ?

**Review by mbayopanda — APPROVED** [review]
It works well on my computer

File: client/src/partials/patients/edit/updatePatientGroups.modal.js (hunk 3 above)
Comment (jniles) [reply client/src/partials/patients/edit/updatePatientGroups.modal.js]: Sure, I'll make the change.

**Comment by jniles** [comment]
@mbayopanda, requested change has landed.  Let me know if you have any further comments.

**Pull request merged**